vault/expiration.go

```diff
@@ -279,63 +290,139 @@ func (m *ExpirationManager) Restore(errorFunc func(), loadDelay time.Duration) (
 	}
 	m.logger.Debug("expiration: leases collected", "num_existing", len(existing))
 
-	// Restore each key by pulling from the result chan
-	for i, leaseID := range existing {
-		if i > 0 && i%500 == 0 {
-			m.logger.Trace("expiration: leases loading", "progress", i)
+	// Make the channels used for the worker pool
+	broker := make(chan string)
+	quit := make(chan bool)
+	// Buffer these channels to prevent deadlocks
+	errs := make(chan error, len(existing))
+	result := make(chan *leaseEntry, len(existing))
+
+	// Use a wait group
+	wg := &sync.WaitGroup{}
+
+	// Create 64 workers to distribute work to
+	for i := 0; i < consts.ExpirationRestoreWorkerCount; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				select {
+				case leaseID, ok := <-broker:
+					// broker has been closed, we are done
+					if !ok {
+						return
+					}
+
+					// If we loaded a lease elsewhere in the expiration manager during
+					// restore, send a nil entry to the loop waiting for all
+					// existing to be processed
+					if _, ok := m.restoreLoaded.Load(leaseID); ok {
+						result <- nil
+						continue
+					}
+
+					// Useful for testing to add latency to all load requests
+					if loadDelay > 0 {
+						time.Sleep(loadDelay)
+					}
+
+					m.lockLease(leaseID)
+					le, err := m.loadEntryInternal(leaseID, false)
+					m.unlockLease(leaseID)
+					if err != nil {
+						errs <- err
+						continue
+					}
+
+					// Write results out to the result channel
+					result <- le
+
+				// quit early
+				case <-quit:
+					return
+				}
+			}
+		}()
+	}
+
+	// Distribute the collected keys to the workers in a go routine
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for i, leaseID := range existing {
+			if i > 0 && i%500 == 0 {
+				m.logger.Trace("expiration: leases loading", "progress", i)
+			}
+
+			select {
+			case <-quit:
+				return
+
+			default:
+				broker <- leaseID
+			}
 		}
 
-		if err := m.loadAndRestoreLease(leaseID); err != nil {
+		// Close the broker, causing worker routines to exit
+		close(broker)
+	}()
+
+	// Restore each key by pulling from the result chan
+	for i := 0; i < len(existing); i++ {
+		select {
+		case err := <-errs:
+			// Close all go routines
+			close(quit)
+
 			return err
-		}
 
-		// This exists for test only to add latency to load calls
-		if loadDelay > 0 {
-			time.Sleep(loadDelay)
+		case le := <-result:
+			// If there is no entry, nothing to restore
+			if le == nil {
+				continue
+			}
+
+			m.lockLease(le.LeaseID)
+			m.restoreLease(le, true)
```

Comment:
Doesn't loadEntryInternal already call restoreLease?

Reply:
OK, I see that it doesn't, because of the true/false, but I'm not sure we are actually gaining anything by sending it from the one goroutine to the other at this point. If the operation that is slow is loading from disk/network, why skip the restoreLease call -- we've already done the load, why not just restoreLease right then and skip a bunch of this coordination?

Reply:
In addition, above, we are already holding the lease lock (although as i stated I think it should be moved up) so if we don't give it up we don't have to worry about as many places where we might have a race.